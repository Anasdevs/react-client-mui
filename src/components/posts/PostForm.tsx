import { SubmitHandler, useForm } from "react-hook-form";
import { yupResolver } from "@hookform/resolvers/yup";
import * as yup from "yup";
import { useNavigate } from "react-router-dom";
import {
    Alert,
    Box,
    Button,
    FormControl,
    FormHelperText,
    InputLabel,
    MenuItem,
    Select,
    Stack,
    TextField,
} from "@mui/material";
import useAuth from "../../hooks/useAuth";
import { useState } from "react";
import { AxiosError, AxiosResponse, isAxiosError } from "axios";
import { PostFormValues } from "../../types";

const validationSchema = yup.object().shape({
    title: yup.string().max(60).required(),
    text: yup.string().min(50).required(),
    native_text: yup.string(),
    language: yup.string().required(),
    gender_of_narration: yup.string(),
    permission: yup.string(),
    // tags: yup.string(),
});

interface Props {
    post: PostFormValues | undefined;
    onSubmit: (data: PostFormValues) => Promise<AxiosResponse>;
    onDiscard: () => void;
}

const PostForm = ({ post, onSubmit, onDiscard }: Props) => {
    const navigate = useNavigate();
    const isEditing = post !== undefined && Object.keys(post).length > 0;
    const [errMsg, setErrMsg] = useState("");
    const {
        register,
        handleSubmit,
        formState: { errors },
        reset,
    } = useForm<PostFormValues>({
        resolver: yupResolver(validationSchema),
    });

    const authContext = useAuth();
    if (!authContext) {
        throw new Error("authContext missing");
    }

    const { currentUser } = authContext;

    const onSubmitHandler: SubmitHandler<PostFormValues> = async (data) => {
        try {
            const response = await onSubmit(data);
            reset();

            if (!isEditing) {
                const { slug } = response.data;
                navigate(`/journals/${slug}`);
            }
        } catch (err) {
            const error = err as Error | AxiosError;

            if (!isAxiosError(error)) {
                setErrMsg("An error has occured.");
            } else {
                if (!error?.response) {
                    setErrMsg("No server response");
                } else if (error.response?.status === 403) {
                    setErrMsg(error.response?.data?.detail);
                } else {
                    setErrMsg("Action failed. Try again.");
                }
            }
        }
    };

    return (
        <form method="post" onSubmit={handleSubmit(onSubmitHandler)}>
            <Stack spacing={3} mb={3}>
                <TextField
                    required
                    {...register("title")}
                    label="Title"
                    error={!!errors.title}
                    helperText={errors.title && String(errors.title?.message)}
                    defaultValue={post?.title}
                />
                <TextField
                    multiline
                    required
                    {...register("text")}
                    label="Write in your target language"
                    error={!!errors.text}
                    helperText={errors.text && String(errors.text?.message)}
                    defaultValue={post?.text || ""}
                />
                <TextField
                    multiline
                    {...register("native_text")}
                    label="Notes"
                    error={!!errors.native_text}
                    helperText={
                        errors.native_text &&
                        String(errors.native_text?.message)
                    }
                    defaultValue={post?.native_text || ""}
                />
                <FormControl error={!!errors.language} fullWidth required>
                    <InputLabel id="language">Language</InputLabel>
                    <Select
                        {...register("language")}
                        label="Language"
                        defaultValue={post?.language || ""}
                    >
                        {currentUser?.get_studying_languages?.map((lang) => (
                            <MenuItem key={lang.code} value={lang.code}>
                                {lang.en_name}
                            </MenuItem>
                        ))}
                    </Select>
                    {errors.language && (
                        <FormHelperText>
                            {String(errors.language.message)}
                        </FormHelperText>
                    )}
                </FormControl>
                <FormControl
                    error={!!errors.gender_of_narration}
                    fullWidth
                    required
                >
                    <InputLabel id="gender">Gender of Narration</InputLabel>
                    <Select
                        {...register("gender_of_narration")}
                        label="Gender of Narration"
                        defaultValue={post?.gender_of_narration || "U"}
                    >
                        <MenuItem value="M">Male</MenuItem>
                        <MenuItem value="F">Female</MenuItem>
                        <MenuItem value="O">Other</MenuItem>
                        <MenuItem value="U">Prefer not to say</MenuItem>
                    </Select>
                    {errors.gender_of_narration && (
                        <FormHelperText>
                            {String(errors.gender_of_narration.message)}
                        </FormHelperText>
                    )}
                </FormControl>
                <FormControl error={!!errors.permission} fullWidth required>
                    <InputLabel id="visibility">Visibility</InputLabel>
                    <Select
                        {...register("permission")}
                        label="Visibility"
                        defaultValue={post?.permission || "public"}
                    >
                        <MenuItem value="public">Viewable by everyone</MenuItem>
                        <MenuItem value="member">
                            Viewable only by registered members
                        </MenuItem>
                    </Select>
                    {errors.permission && (
                        <FormHelperText>
                            {String(errors.permission.message)}
                        </FormHelperText>
                    )}
                </FormControl>
            </Stack>

            {errMsg && (
                <Alert severity="error" sx={{ mb: 3 }}>
                    {errMsg}
                </Alert>
            )}

            <Box display="flex" justifyContent="end" gap={1}>
                <Button
                    variant="outlined"
                    color="error"
                    onClick={() => {
                        reset();
                        onDiscard();
                    }}
                >
                    {isEditing ? "Discard changes" : "Cancel"}
                </Button>
                <Button type="submit" variant="contained">
                    {isEditing ? "Save" : "Submit"}
                </Button>
            </Box>
        </form>
    );
};

export default PostForm;
