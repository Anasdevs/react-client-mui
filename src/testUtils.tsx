import AuthContext from "./context/AuthProvider";

// TODO: populate demo user

const demoUser = {
    username: "testuser",
};

const TestAuthProvider = ({
    children,
    accessToken = "123",
    refreshToken = "123",
    setAccessToken = () => {
        /** */
    },
    setRefreshToken = () => {
        /** */
    },
    currentUser = demoUser,
    setCurrentUser = () => {
        /** */
    },
    isAuthenticated = true,
    userInfoLoaded = true,
    logout = () => {
        /** */
    },
}) => (
    <AuthContext.Provider
        value={{
            accessToken,
            setAccessToken,
            refreshToken,
            setRefreshToken,
            currentUser,
            setCurrentUser,
            isAuthenticated,
            userInfoLoaded,
            logout,
        }}
    >
        {children}
    </AuthContext.Provider>
);

export { TestAuthProvider };
