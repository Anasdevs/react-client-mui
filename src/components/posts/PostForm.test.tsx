import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import PostForm from "./PostForm";
import { TestAuthProvider } from "../../testUtils";

// TODO: woefully under-tested

const handleSubmit = () => {};

it("renders without crashing", () => {
    render(
        <BrowserRouter>
            <TestAuthProvider>
                <PostForm post={undefined} onSubmit={handleSubmit} />
            </TestAuthProvider>
        </BrowserRouter>,
    );
});

it("matches snapshot", () => {
    const { container } = render(
        <BrowserRouter>
            <TestAuthProvider>
                <PostForm post={undefined} onSubmit={handleSubmit} />
            </TestAuthProvider>
        </BrowserRouter>,
    );
    expect(container).toMatchSnapshot();
});
