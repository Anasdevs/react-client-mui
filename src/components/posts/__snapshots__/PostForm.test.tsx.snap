// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`matches snapshot 1`] = `
<div>
  <form
    method="post"
  >
    <div
      class="MuiStack-root css-1cn8xdu-MuiStack-root"
    >
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r3:"
          id=":r3:-label"
        >
          Title
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r3:"
            name="title"
            required=""
            type="text"
            value=""
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Title
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r4:"
          id=":r4:-label"
        >
          Write in your target language
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-dpjnhs-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r4:"
            name="text"
            required=""
            rows="8"
            style="height: 0px; overflow: hidden;"
          />
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
            tabindex="-1"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Write in your target language
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          for=":r5:"
          id=":r5:-label"
        >
          Notes
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-multiline css-dpjnhs-MuiInputBase-root-MuiOutlinedInput-root"
        >
          <textarea
            aria-invalid="false"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            id=":r5:"
            name="native_text"
            rows="8"
            style="height: 0px; overflow: hidden;"
          />
          <textarea
            aria-hidden="true"
            class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputMultiline css-1sqnrkk-MuiInputBase-input-MuiOutlinedInput-input"
            readonly=""
            style="visibility: hidden; position: absolute; overflow: hidden; height: 0px; top: 0px; left: 0px; transform: translateZ(0); padding: 0px; width: 100%;"
            tabindex="-1"
          />
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Notes
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="false"
          id="language"
        >
          Language
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="mui-component-select-language"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-component-select-language"
            role="button"
            tabindex="0"
          >
            <span
              class="notranslate"
            >
              ​
            </span>
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            name="language"
            required=""
            tabindex="-1"
            value=""
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-yjsfm1"
            >
              <span>
                Language
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          id="gender"
        >
          Gender of Narration
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="mui-component-select-gender_of_narration"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-component-select-gender_of_narration"
            role="button"
            tabindex="0"
          >
            Prefer not to say
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            name="gender_of_narration"
            required=""
            tabindex="-1"
            value="U"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Gender of Narration
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
      <div
        class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
      >
        <label
          class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
          data-shrink="true"
          id="visibility"
        >
          Visibility
          <span
            aria-hidden="true"
            class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
        >
          <div
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="mui-component-select-permission"
            class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
            id="mui-component-select-permission"
            role="button"
            tabindex="0"
          >
            Viewable by everyone
          </div>
          <input
            aria-hidden="true"
            aria-invalid="false"
            class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
            name="permission"
            required=""
            tabindex="-1"
            value="public"
          />
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
            data-testid="ArrowDropDownIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
          >
            <legend
              class="css-14lo706"
            >
              <span>
                Visibility
                 
                *
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="MuiBox-root css-qw1cjb"
    >
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedError MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-180lf2n-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Cancel
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-sghohy-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="submit"
      >
        Submit
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </form>
</div>
`;
